.pre-commit-config.yaml

```diff
@@ -0,0 +1,57 @@
+repos:
+
+  - repo: local
+    hooks:
+
+      # Custom
+      - id: Validate CircleCI
+        name: Validate CircleCI Config
+        entry: ./scripts/hooks/validate_circleci_config.sh
+        language: script
+
+  - repo: https://github.com/pre-commit/pre-commit-hooks
+    rev: v2.5.0
+    hooks:
+
+      # Git
+      - id: no-commit-to-branch
+        args: ['--branch', 'master']
+      - id: forbid-new-submodules
+
+      # Files
+      - id: check-byte-order-marker
+      - id: check-executables-have-shebangs
+      - id: check-added-large-files
+      - id: check-symlinks
```

Comment:
Any thoughts on `end-of-file-fixer`?

I tried running it locally and ended up with:
```
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git checkout -- <file>..." to discard changes in working directory)

	modified:   .bandit
	modified:   .bumpversion.cfg
	modified:   .circleci/validate-hook.sh
	modified:   .github/labeler.yml
	modified:   .pre-commit-config.yaml
	modified:   CODE_OF_CONDUCT.md
	modified:   deploy/ansible/inventory/ec2.py
	modified:   deploy/ansible/ursulas/rename_configs.yml
	modified:   deploy/ansible/variables.template.yml
	modified:   deploy/ansible/worker/get_workers_status.yml
	modified:   deploy/ansible/worker/include/check_running_ursula.yml
	modified:   deploy/ansible/worker/include/run_ursula.yml
	modified:   deploy/ansible/worker/include/setup_docker.yml
	modified:   deploy/ansible/worker/include/setup_user.yml
	modified:   deploy/ansible/worker/include/update_existing_ursula.yml
	modified:   deploy/ansible/worker/setup_remote_workers.yml
	modified:   deploy/ansible/worker/update_remote_workers.yml
	modified:   deploy/geth/initialize_blockchain.sh
	modified:   deploy/geth/launch_bootnode.sh
	modified:   deploy/static-nodes.json
	modified:   docs/Makefile
	modified:   docs/source/api/config.rst
	modified:   docs/source/api/crypto.rst
	modified:   docs/source/api/keyring.rst
	modified:   docs/source/api/network.rst
	modified:   docs/source/demos/heartbeat_demo.rst
	modified:   docs/source/guides/contribution_guide.rst
	modified:   docs/source/guides/deployment_guide.rst
	modified:   docs/source/guides/development/development.rst
	modified:   docs/source/guides/federated_testnet_guide.rst
	modified:   docs/source/guides/network_node/network_node.rst
	modified:   docs/source/support/node_providers.rst
	modified:   examples/heartbeat_demo/alicia.py
	modified:   mypy.ini
	modified:   nucypher/blockchain/eth/contract_registry/cassandra/contract_registry.json
	modified:   nucypher/blockchain/eth/contract_registry/frances/contract_registry.json
	modified:   nucypher/blockchain/eth/contract_registry/gemini/contract_registry.json
	modified:   nucypher/blockchain/eth/contract_registry/miranda/README.md
	modified:   nucypher/blockchain/eth/contract_registry/miranda/contract_registry.json
	modified:   nucypher/blockchain/eth/contract_registry/miranda/individual_allocation_ABI.json
	modified:   nucypher/blockchain/eth/decorators.py
	modified:   nucypher/blockchain/eth/events.py
	modified:   nucypher/blockchain/eth/multisig.py
	modified:   nucypher/blockchain/eth/sol/source/zeppelin/utils/Address.sol
	modified:   nucypher/characters/base.py
	modified:   nucypher/characters/control/specifications/fields/treasuremap.py
	modified:   nucypher/cli/options.py
	modified:   nucypher/crypto/splitters.py
	modified:   nucypher/datastore/threading.py
	modified:   nucypher/network/nicknames/zodiac.json
	modified:   nucypher/network/templates/basic_status.j2
	modified:   nucypher/network/trackers.py
	modified:   nucypher/utilities/sandbox/blockchain.py
	modified:   nucypher/utilities/sandbox/constants.py
	modified:   nucypher/utilities/sandbox/middleware.py
	modified:   pytest.ini
	modified:   readthedocs.yml
	modified:   scripts/circle/logOutput.sh
	modified:   scripts/circle/rebuild_pipenv.sh
	modified:   scripts/local_fleet/run_local_fleet.sh
	modified:   scripts/local_fleet/run_local_ursula_fleet.py
	modified:   tests/blockchain/eth/contracts/lib/test_reencryption_validator.py
	modified:   tests/blockchain/eth/entities/actors/test_multisig_actors.py
	modified:   tests/blockchain/eth/entities/deployers/test_deploy_idle_network.py
	modified:   tests/blockchain/eth/entities/deployers/test_multisig_deployer.py
	modified:   tests/blockchain/eth/interfaces/test_chains.py
	modified:   tests/blockchain/eth/interfaces/test_registry.py
	modified:   tests/characters/control/federated/test_rpc_control_federated.py
	modified:   tests/cli/test_deploy_commands.py
	modified:   tests/cli/test_multisig_cli.py
	modified:   tests/cli/ursula/test_federated_ursula.py
	modified:   tests/config/test_base_configuration.py
	modified:   tests/conftest.py
	modified:   tests/crypto/test_bytestring_types.py
	modified:   tests/fixtures.py
```

Reply:
Heh - Yeah.  We have many files whiteout proper white space at the end!  I think it will be better to introduce more of these changes incrementally.  Feel free to open a PR  with the white space fixes :-)

Reply:
👍 - Filed #1920 .